strawberryfields/apps/similarity.py

```diff
@@ -306,33 +320,148 @@ def event_cardinality(photon_number: int, max_count_per_mode: int, modes: int) -
     return cardinality
 
 
-def prob_orbit_mc(
-    graph: nx.Graph, orbit: list, n_mean: float = 5, samples: int = 1000, loss: float = 0.0
+def _prob_orbit_exact(
+    graph: nx.Graph,
+    orbit: list,
+    n_mean: float = 5,
+    loss: float = 0.0
 ) -> float:
-    r"""Gives a Monte Carlo estimate of the GBS probability of a given orbit according to the input
+    r"""Gives the exact GBS probability of a given orbit according to the input
     graph.
 
-    To make this estimate, several samples from the orbit are drawn uniformly at random using
-    :func:`orbit_to_sample`. The GBS probabilities of these samples are then calculated and the
-    sum is used to create an estimate of the orbit probability.
+    The exact probability of an orbit is made up of the sum of probabilities of
+    all possible GBS output patterns that belong to it:
+
+    .. math::
+       P(O) = \sum_{n \in O} P(n) = \frac{{\rm haf}(A_{n})^2}{n! \sqrt{\det(Q)}}
+
+    where :math:`n` represents one GBS output pattern, :math:`A_{n}` is its induced
+    sub-matrix and :math:`Q` is the Q-matrix obtained from the adjacency matrix.
+
+    **Example usage:**
+
+    >>> graph = nx.complete_graph(8)
+    >>> _prob_orbit_exact(graph, [2, 1, 1])
+    0.03744
+
+    Args:
+        graph (nx.Graph): input graph encoded in the GBS device
+        orbit (list[int]): orbit for which to calculate the probability
+        n_mean (float): total mean photon number of the GBS device
+        loss (float): fraction of photons lost in GBS
+
+    Returns:
+        float: exact orbit probability
+    """
+
+    if n_mean < 0:
+        raise ValueError("Mean photon number must be non-negative")
+    if not 0 <= loss <= 1:
+        raise ValueError("Loss parameter must take a value between zero and one")
+
+    modes = graph.order()
+    photons = sum(orbit)
+    A = nx.to_numpy_array(graph)
+    mean_photon_per_mode = n_mean / float(modes)
+
+    p = sf.Program(modes)
+
+    with p.context as q:
+        sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
+
+        if loss:
+            for _q in q:
+                sf.ops.LossChannel(1 - loss) | _q
+
+    eng = sf.LocalEngine(backend="gaussian")
+    result = eng.run(p)

```

Comment:
This sort of logic seems to crop up a few times now. I'd recommend factoring out to make a private function that encodes the problem into SF:
```python
def _get_state(graph, n_mean, loss):
    modes = graph.order()
    photons = sum(orbit)
    A = nx.to_numpy_array(graph)
    mean_photon_per_mode = n_mean / float(modes)
    
    p = sf.Program(modes)
    
    with p.context as q:
        sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
    
        if loss:
            for _q in q:
                sf.ops.LossChannel(1 - loss) | _q
    
    eng = sf.LocalEngine(backend="gaussian")
    return eng.run(p).state
```
This would help separate the two functionalities of (i) embedding into SF, (ii) using the SF state to calculate the probability.